File: sdks/go/lunora/key.go
package lunora

import (
	"bytes"
	"encoding/json"
	"math"
	"sort"
	"strconv"
	"strings"
)

// StableStringify renders a pure-JSON tree canonically: object keys sorted at
// every depth by code point, arrays keeping their order, null fields kept, and
// Undefined object fields dropped.
//
// It runs on the OUTPUT of EncodeWire, so it only ever sees nil, bool, float64,
// string, []any and map[string]any. Two arg records that differ only in key
// insertion order therefore collapse to one key — which is the point: this is
// what de-duplicates subscriptions.
func StableStringify(value any) string {
	var builder strings.Builder

	writeStable(&builder, value)

	return builder.String()
}

// StableWireKey is the stable cache/dedup key for value.
func StableWireKey(value any) (string, error) {
	encoded, err := EncodeWire(value)
	if err != nil {
		return "", err
	}

	return StableStringify(encoded), nil
}

func writeStable(builder *strings.Builder, value any) {
	switch typed := value.(type) {
	case nil:
		builder.WriteString("null")
	case undefinedType:
		// Only reachable in an array position; an object field was dropped by
		// the caller. JSON has no undefined, and null is what JSON.stringify
		// would have produced here.
		builder.WriteString("null")
	case bool:
		builder.WriteString(strconv.FormatBool(typed))
	case string:
		builder.WriteString(jsonString(typed))
	case float64:
		builder.WriteString(formatNumber(typed))
	case []any:
		builder.WriteByte('[')

		for index, item := range typed {
			if index > 0 {
				builder.WriteByte(',')
			}

			writeStable(builder, item)
		}

		builder.WriteByte(']')
	case map[string]any:
		writeStableObject(builder, typed)
	default:
		// EncodeWire has already rejected anything else; emit null rather than
		// panicking on a caller that reached here with a raw value.
		builder.WriteString("null")
	}
}

func writeStableObject(builder *strings.Builder, value map[string]any) {
	keys := make([]string, 0, len(value))

	for key := range value {
		if value[key] == any(Undefined) {
			continue
		}

		keys = append(keys, key)
	}

	// Code-point order, matching JavaScript's default string comparison. Go's
	// string < is byte-wise over UTF-8, which is the same ordering.
	sort.Strings(keys)
	builder.WriteByte('{')

	for index, key := range keys {
		if index > 0 {
			builder.WriteByte(',')
		}

		builder.WriteString(jsonString(key))
		builder.WriteByte(':')
		writeStable(builder, value[key])
	}

	builder.WriteByte('}')
}

// formatNumber matches JSON.stringify: an integral float drops its decimal.
func formatNumber(value float64) string {
	if math.IsNaN(value) || math.IsInf(value, 0) {
		// Unreachable via EncodeWire (both are tagged before this runs), but
		// JSON.stringify would emit null and so do we.
		return "null"
	}

	if value == math.Trunc(value) && math.Abs(value) < 1e21 {
		return strconv.FormatFloat(value, 'f', -1, 64)
	}

	return strconv.FormatFloat(value, 'g', -1, 64)
}

// jsonString quotes a string the way JSON.stringify does. Go's encoder escapes
// <, > and & for HTML safety by default, which JavaScript does not — that would
// produce a different key for the same args, so it is turned off.
func jsonString(value string) string {
	var buffer bytes.Buffer

	encoder := json.NewEncoder(&buffer)
	encoder.SetEscapeHTML(false)

	if err := encoder.Encode(value); err != nil {
		return `""`
	}

	return strings.TrimRight(buffer.String(), "\n")
}

File: sdks/ruby/lib/lunora/key.rb
# frozen_string_literal: true

require "json"

require_relative "wire"

module Lunora
  module_function

  # Canonical JSON encoding of a pure-JSON tree: object keys sorted at every
  # depth, arrays keeping their order, null fields kept, UNDEFINED object fields
  # dropped.
  #
  # Runs on the OUTPUT of +encode_wire+, so it only ever sees nil/true/false/
  # Integer/Float/String/Array/Hash. Two arg records differing only in key
  # insertion order collapse to one key — which is the point: this is what
  # de-duplicates subscriptions, and it is compared verbatim against a key
  # produced by the reference TypeScript client.
  def stable_stringify(value)
    case value
    when UNDEFINED, nil then "null"
    when true then "true"
    when false then "false"
    when ::Integer then value.to_s
    when ::Float then format_number(value)
    when ::String then json_string(value)
    when ::Array then "[#{value.map { |item| stable_stringify(item) }.join(",")}]"
    when ::Hash then stable_object(value)
    else "null"
    end
  end

  # The stable cache/dedup key for +value+.
  def stable_wire_key(value)
    stable_stringify(encode_wire(value))
  end

  def stable_object(value)
    pairs = value.reject { |_key, item| item.equal?(UNDEFINED) }
    # JavaScript compares strings by UTF-16 code unit. Ruby's String <=> is
    # byte-wise over UTF-8, which agrees inside the BMP but not above it: an
    # astral character is its high surrogate (0xD83D) as UTF-16 and 0xF0.. as
    # UTF-8, so it sorts before U+FFFD there and after it here. Encoding to
    # UTF-16BE before comparing reproduces JavaScript's ordering exactly.
    sorted = pairs.sort_by { |key, _item| key.to_s.encode(::Encoding::UTF_16BE, invalid: :replace, undef: :replace).b }
    "{#{sorted.map { |key, item| "#{json_string(key.to_s)}:#{stable_stringify(item)}" }.join(",")}}"
  end

  # Renders a float exactly as +String(v)+ does in JavaScript, which is what
  # JSON.stringify emits for a finite number (ECMA-262 Number::toString).
  #
  # The DIGITS come from Float#to_s, which is already the shortest string that
  # reads back as the same double; only their LAYOUT differs (Ruby writes
  # "1.0e-05" and switches to exponent form at 1e16). Re-laying those digits is
  # the whole job. Searching a fixed number of decimal places instead capped the
  # output at 20 places, so three adjacent doubles near -6.07e-6 all keyed as
  # "-0.00000607387560669604" and one subscription received another's frames.
  def format_number(value)
    return "null" if value.nan? || value.infinite?

    # A negative zero keys as "-0": the sign is taken from the bits, not from
    # a comparison, which -0.0 < 0 would get wrong.
    sign = value.negative? || (value.zero? && (1.0 / value).negative?) ? "-" : ""
    digits, point = shortest_digits(value.abs)
    return "#{sign}0" if digits.empty?

    sign + ecma_layout(digits, point)
  end

  # The shortest round-trip digits of a finite non-negative double, and n such
  # that the value is 0.<digits> x 10^n.
  def shortest_digits(magnitude)
    mantissa, exponent = magnitude.to_s.split("e")
    whole, fraction = mantissa.split(".")
    digits = whole + fraction.to_s
    point = whole.length + exponent.to_i
    leading = digits[/\A0*/].length

    [digits[leading..].sub(/0+\z/, ""), point - leading]
  end

  # ECMA-262 Number::toString, steps for k digits and decimal exponent n.
  def ecma_layout(digits, point)
    count = digits.length
    return digits + ("0" * (point - count)) if point.between?(count, 21)
    return "#{digits[0, point]}.#{digits[point..]}" if point.positive? && point <= 21
    return "0.#{"0" * -point}#{digits}" if point > -6 && point <= 0

    exponent = point - 1
    mantissa = count == 1 ? digits : "#{digits[0]}.#{digits[1..]}"
    "#{mantissa}e#{exponent.negative? ? "-" : "+"}#{exponent.abs}"
  end

  # Quotes a string the way JSON.stringify does. Ruby's JSON generator already
  # leaves <, > and & raw and does not escape U+2028/U+2029, so it matches
  # without the adjustments the Go port needs.
  def json_string(value)
    ::JSON.generate(value.to_s, quirks_mode: true)
  end
end
